File: 1208.js
const fs = require('fs');

function solve() {
	let input = fs.readFileSync('/dev/stdin').toString().split('\n');
	let inputIdx = 0;

	const [N, S] = input[inputIdx++].split(' ').map((v) => +v);

	const arr = input[inputIdx++].split(' ').map((v) => +v);

	const extend = (seq) => {
		let ret = [0];
		for (const s of seq) {
			const tmp = ret.map((v) => v + s);
			ret = ret.concat(tmp);
		}
		return ret;
	};

	const counter = (seq) => {
		const count = {};
		for (const s of seq) {
			count.hasOwnProperty(s) ? (count[s] += 1) : (count[s] = 1);
		}
		return count;
	};

	const idx = Math.floor(N / 2);
	const leftArr = extend(arr.slice(0, idx));
	const rightArr = extend(arr.slice(idx));
	const rightCounter = counter(rightArr);

	let ans = 0;
	for (const v of leftArr) {
		const remain = S - v;
		if (rightCounter.hasOwnProperty(remain)) ans += rightCounter[remain];
	}

	return S === 0 ? ans - 1 : ans;
}

console.log(solve());

File: 1208.py
from sys import stdin

input = stdin.readline


def solve():
    N, S = map(int, input().split())
    arr = [*map(int, input().split())]

    def extend(brr):
        ret = [0]
        for b in brr:
            tmp = [b + c for c in ret]
            ret += tmp
        return ret

    def count_sub(brr):
        dic = {0: 1}
        for b in brr:
            tmp = {}
            for v in dic:
                tmp[b + v] = dic[v]
            for v in tmp:
                if v in dic:
                    dic[v] += tmp[v]
                else:
                    dic[v] = tmp[v]

        return dic

    idx = N//2
    left_arr = extend(arr[:idx])
    right_dic = count_sub(arr[idx:])

    ans = 0
    for v in left_arr:
        remain = S - v
        if remain in right_dic:
            ans += right_dic[remain]

    return ans if S != 0 else ans - 1


if __name__ == '__main__':
    print(solve())
